src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -1555,629 +1559,676 @@ private void saveSecondaryInterfaces(final List<String> newCmdComps, final List<
-
+		LOGGER.log(Level.INFO, "Beginning GCM handshake");
+		// send first part of handshake - contains userid
+		Request wrapper = buildClientHandshakeGCMRequest(userid, handshake);
+
+		CompletionStage<SQLResult<Void>> GCMHandshakeCompletionStage = sendRequest(
+			wrapper,
+			wireMessage -> {
+				final ClientWireProtocol.ClientConnectionGCMResponse.Builder ccr = ClientWireProtocol.ClientConnectionGCMResponse.newBuilder();
+				ccr.mergeFrom(new ByteBufInputStream(wireMessage));
+				return ccr;
+			},
+			this::handleWarning,
+			ClientWireProtocol.ClientConnectionGCMResponse.Builder::getResponse
+		).thenApply(ccrOrEx -> ccrOrEx.mapSuccess(ccr -> {
+			// Handle first handshake response
 			final ByteString ivString = ccr.getIv();
-			byte[] key;
-			byte[] macKey;
-			String myPubKey;
+			ClientHandshakeCrypto crypto;
 
 			try
 			{
-				String keySpec = ccr.getPubKey();
-				keySpec = keySpec.replace("-----BEGIN PUBLIC KEY-----\n", "");
-				keySpec = keySpec.replace("-----END PUBLIC KEY-----\n", "");
-				final byte[] keyBytes = Base64.getMimeDecoder().decode(keySpec.getBytes(StandardCharsets.UTF_8));
-				final X509EncodedKeySpec x509keySpec = new X509EncodedKeySpec(keyBytes);
-				final KeyFactory keyFact = KeyFactory.getInstance("DH");
-				final DHPublicKey pubKey = (DHPublicKey) keyFact.generatePublic(x509keySpec);
-				final DHParameterSpec params = pubKey.getParams();
-
-				final KeyPairGenerator keyGen = KeyPairGenerator.getInstance("DH");
-				keyGen.initialize(params);
-				final KeyPair kp = keyGen.generateKeyPair();
-
-				final KeyAgreement ka = KeyAgreement.getInstance("DiffieHellman");
-				ka.init(kp.getPrivate());
-				ka.doPhase(pubKey, true);
-				final byte[] secret = ka.generateSecret();
-
-				final byte[] buffer = new byte[5 + secret.length];
-				buffer[0] = (byte) ((secret.length & 0xff000000) >> 24);
-				buffer[1] = (byte) ((secret.length & 0xff0000) >> 16);
-				buffer[2] = (byte) ((secret.length & 0xff00) >> 8);
-				buffer[3] = (byte) (secret.length & 0xff);
-				System.arraycopy(secret, 0, buffer, 5, secret.length);
-
-				buffer[4] = 0x00;
-				MessageDigest sha = MessageDigest.getInstance("SHA-256");
-				key = sha.digest(buffer);
-
-				buffer[4] = 0x01;
-				sha = MessageDigest.getInstance("SHA-256");
-				macKey = sha.digest(buffer);
-
-				final PublicKey clientPub = kp.getPublic();
-				myPubKey = "-----BEGIN PUBLIC KEY-----\n" + Base64.getMimeEncoder().encodeToString(clientPub.getEncoded()) + "\n-----END PUBLIC KEY-----\n";
-			}
-			catch(RuntimeException e)
-			{
-				LOGGER.log(Level.WARNING, String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
-				throw e;
+				crypto = clientHandshakePublicKeyCrypto(ccr.getPubKey());
 			}
 			catch (final Exception e)
 			{
 				LOGGER.log(Level.WARNING, String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
-				throw e;
+				return e;
```

Comment:
does this compile?

I would expect the return type of `SQLResult<T>#mapSuccess` to be `SQLResult<U>` where `U` is `EncryptedHandshake` in this case.

Maybe this?
```suggestion
				return SQLResult.failOrMarshal(e);
```

Reply:
Yes it compiles and I ran the unit tests. Looks like `mapSuccess` wraps the return value inside a SQLResult. 

Both `return encryptedHand` and `return e` are inside the lambda function passed into `mapSuccess`

Reply:
If we do `return SQLResult.failOrMarshal(e)`, then `mapSuccess` would return type `SQLResult<SQLResult<Object>>` instead of the desired `SQLResult<Object>`

Reply:
For instance, this doesn't compile for me:
```
@Test
public void testMapSuccessToException() throws Exception {
    SQLException ex = new SQLException("My exception");

    SQLResult<Void> result = SQLResult
        .voidResult()
        .mapSuccess(o -> {
            return ex;
        });

    assertTrue(result.isException());
    assertFalse(result.isSuccess());
    assertTrue("Expected the unmarshaled exception instance", ex == result.getException());
}
```

Reply:
U is Object. I end up casting in the logic below which isn't great. We can switch back to `thenCompose` + `mapSuccessAsync` which allows us to guarantee a return type of `SQLResult<EncryptedHandshake>`?

Reply:
```
.thenCompose(ccrOrEx -> ccrOrEx.mapSuccessAsync(ccr -> {
			// Handle first handshake response
			final ByteString ivString = ccr.getIv();
			ClientHandshakeCrypto crypto;

			try
			{
				crypto = clientHandshakePublicKeyCrypto(ccr.getPubKey());
			}
			catch (final Exception e)
			{
				LOGGER.log(Level.WARNING, String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
				return SQLResult.failOrMarshalAsync(e);
			}

			// We are using a 16 byte authentication tag on the server side. 16 * 8 = 128.
			final byte[] iv = ivString.toByteArray();
			final GCMParameterSpec gps = new GCMParameterSpec(128, iv);
			EncryptedHandshake encryptedHand;
			try {
				encryptedHand = clientHandshakeEncryption(crypto, gps, "AES/GCM/NoPadding");
			} catch (Exception e) {
				return SQLResult.failOrMarshalAsync(e);
			}

			return SQLResult.successAsync(encryptedHand);
		}))
```

Reply:
Ohhh I think I missed that part, hmmm...

I'd prefer _not_ to use `Object` if possible. A ton of legacy code does that, making it hard to do compile-time checks, for instance, ensuring that all possible variants of Object are handled, so I've been _trying_ to refactor that code in a type-safe manner. There's a terrible one in `XGStatement#sendAndReceive`.

Is there a common interface you can extract from the set of non-exceptional types that you expect?

Reply:
I think the issue here is that `mapSuccess` wraps the return value of the lambda in SQLResult before returning.

Whereas `mapSuccessAsync` simply returns the return value of the lambda

Reply:
Oh, I see what you're trying to do... what you _really_ want here is "flatMap". Common in monadic/functional interfaces and something I didn't have a use case for yet so I didn't add this functionality to `*Result`, but this seems like a good use case!

Here is an [example](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#flatMap-java.util.function.Function-) of this pattern in java's `Optional` implementation.

In our case, we want:
```
default <U> SQLResult<U> flatMapSuccess(final Function<? super T, ? extends SQLResult<U>> fn) {
    return fold(fn::apply, SQLResult::fail);
}
```

"flatMapSuccess" to stay consistent with the "success" "fail"/"ex" pattern.

Reply:
> Whereas mapSuccessAsync simply returns the return value of the lambda

Right because that function returns a future that returns the result:
```
default <U> CompletionStage<SQLResult<U>> mapSuccessAsync(final Function<? super T, ? extends CompletionStage<SQLResult<U>>> fn) {
```

Okay, it makes sens why you were trying to do that, I would:
1. Extend the `Result`, `IOResult`, "SQLResult`, and `TransportResult` interfaces with this new method, and then add a test case to CompletableFuturesTest.java